sqldelight-compiler/src/main/kotlin/com/squareup/sqldelight/core/lang/util/TreeUtil.kt

```diff
@@ -161,34 +161,64 @@ fun Collection<SqlDelightQueriesFile>.forInitializationStatements(
   body: (sqlText: String) -> Unit
 ) {
   val views = ArrayList<SqlCreateViewStmt>()
+  val tables = ArrayList<SqlCreateTableStmt>()
   val creators = ArrayList<PsiElement>()
+  val miscellanious = ArrayList<PsiElement>()
 
   forEach { file ->
     file.sqliteStatements()
       .filter { (label, _) -> label.name == null }
       .forEach { (_, sqliteStatement) ->
         when {
+          sqliteStatement.createTableStmt != null -> tables.add(sqliteStatement.createTableStmt!!)
           sqliteStatement.createViewStmt != null -> views.add(sqliteStatement.createViewStmt!!)
           sqliteStatement.createTriggerStmt != null -> creators.add(sqliteStatement.createTriggerStmt!!)
           sqliteStatement.createIndexStmt != null -> creators.add(sqliteStatement.createIndexStmt!!)
-          else -> body(sqliteStatement.rawSqlText())
+          else -> miscellanious.add(sqliteStatement)
         }
       }
   }
 
-  val viewsLeft = views.map { it.viewName.name }.toMutableSet()
-  while (views.isNotEmpty()) {
-    views.removeAll { view ->
-      if (view.compoundSelectStmt!!.findChildrenOfType<SqlTableName>()
-        .any { it.name in viewsLeft }
-      ) {
+  tables.orderStatements(
+    { tableName.name },
+    { table ->
+      table.columnDefList.flatMap {
+        column ->
+        column.columnConstraintList.mapNotNull { it.foreignKeyClause?.foreignTable }
+      }
+    },
+    { name },
+    body,
+  )
+
+  views.orderStatements(
+    { viewName.name },
+    { view: SqlCreateViewStmt -> view.compoundSelectStmt!!.findChildrenOfType<SqlTableName>() },
+    { name },
+    body,
+  )
+
+  creators.forEach { body(it.rawSqlText()) }
+  miscellanious.forEach { body(it.rawSqlText()) }
+}
+
+private fun <T : PsiElement, E : PsiElement> ArrayList<T>.orderStatements(
+  nameSelector: T.() -> String,
```

Comment:
Selectors generally take an argument not have a receiver

Reply:
resolved.